File: spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        rowCount, colCount = len(matrix), len(matrix[0])   
        maxUpperRow, maxLowerRow = 0, rowCount - 1
        maxLeftCol, maxRightCol = 0, colCount - 1

        elementsInSpiralOrder = []

        while len(elementsInSpiralOrder) < rowCount * colCount:
            # move left to right on top
            for col in range(maxLeftCol, maxRightCol + 1):
                elementsInSpiralOrder.append(matrix[maxUpperRow][col])

            # move top to bottom on right (skip top right)
            for row in range(maxUpperRow + 1, maxLowerRow + 1):
                elementsInSpiralOrder.append(matrix[row][maxRightCol])

            # move right to left on bottom (skip if already done when matrix is single row)
            if maxUpperRow != maxLowerRow:
                # skip bottom right
                for col in range(maxRightCol - 1, maxLeftCol - 1, -1):
                    elementsInSpiralOrder.append(matrix[maxLowerRow][col])

            # move bottom to top on left (skip if already done when matrix is single column)
            if maxLeftCol != maxRightCol:
                # skip bottom left, skip top left
                for row in range(maxLowerRow - 1, maxUpperRow, -1):
                    elementsInSpiralOrder.append(matrix[row][maxLeftCol])
                    
            # shift in bounds on all sides
            maxUpperRow += 1
            maxLowerRow -= 1
            maxLeftCol += 1
            maxRightCol -= 1

        return elementsInSpiralOrder            

File: cpp/spiral_matrix.cpp
class Solution {
public:
    // optimal four boundaries approach
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.empty()) {
            return {};
        }

        // initial border boundaries for scanning matrix in spiral
        int totalRows = matrix.size(), totalCols = matrix[0].size();
        int topRow = 0, bottomRow = totalRows - 1;
        int leftCol = 0, rightCol = totalCols - 1;

        vector<int> spiralNums;
        while (spiralNums.size() < totalRows * totalCols) {
            // scan top row of spiral, left to right
            for (int col = leftCol; col <= rightCol; ++col) {
                spiralNums.push_back(matrix[topRow][col]);
            }

            // scan right column of spiral, top to bottom (skip top right)
            for (int row = topRow + 1; row <= bottomRow; ++row) {
                spiralNums.push_back(matrix[row][rightCol]);
            }

            // scan bottom row of spiral, right to left (skip bottom right)
            // do not scan if it's the same top row already scanned left to right
            if (topRow != bottomRow) {
                for (int col = rightCol - 1; col >= leftCol; --col) {
                    spiralNums.push_back(matrix[bottomRow][col]);
                }
            }

            // scan left column of spiral, bottom to top (skip bottom left and top left)
            // do not scan if it's the same right column already scanned top to bottom
            if (leftCol != rightCol) {
                for (int row = bottomRow - 1; row > topRow; --row) {
                    spiralNums.push_back(matrix[row][leftCol]);
                }
            }

            // shift in bounds on all four sides
            ++topRow; --bottomRow;
            ++leftCol; --rightCol;
        }

        return spiralNums;
    }
};
